Add fast boot testing

SebastianCzapla opened the pull request.

SebastianCzapla committed:
dasharo-performance: Add fast boot testing

Signed-off-by: Sebastian Czapla <sebastian.czapla@3mdeb.com>

SebastianCzapla commented:
[fast-boot-tgl.zip](https://github.com/user-attachments/files/19340445/fast-boot-tgl.zip)

miczyg1 commented on dasharo-performance/payload-boot.robot:
Why the file is not named `fast-boot.robot`?

SebastianCzapla replied:
Initially, I went with name "payload-boot.robot" to differentiate it from CBMEM tests, and potentially to make room for future features, but since this feature is limited to NCM, it was not the best choice, so I changed it to suggested name in the recent commit (https://github.com/Dasharo/open-source-firmware-validation/pull/706/commits/b7d12929e5339ec70627efb2ef7f50923dc71cb4).

miczyg1 replied:
> Initially, I went with name "payload-boot.robot" to differentiate it from CBMEM tests, and potentially to make room for future features, but since this feature is limited to NCM, it was not the best choice, so I changed it to suggested name in the recent commit

I don't know where are these assumptions coming from. Fast boot is a generic feature and any platform can enable it... For now it is only NovaCustom at the moment, but it will likely change in the future.

miczyg1 commented on dasharo-performance/payload-boot.robot:
```diff
@@ -0,0 +1,161 @@
+    Login To Linux
+    Switch To Root User
+    Set Fast Boot State    on
+    Set Quiet Boot State    off
```
Why do you touch this option? If you test fast boot, leave others unmodified.

SebastianCzapla replied:
I did it to ensure we are testing only one option at the time, but according to feature description, Fast Boot forces Quiet Boot anyways, so forcing Quiet Boot to be off is redundant; removed it in latest commit (https://github.com/Dasharo/open-source-firmware-validation/pull/706/commits/b7d12929e5339ec70627efb2ef7f50923dc71cb4).

SebastianCzapla replied:
Since with removal of Quiet Boot disabling, the "Set Quiet Boot State" KWD is no longer used, should I remove it as well?

miczyg1 replied:
Obviously

SebastianCzapla replied:
Removed, all changes got squashed into one commit, and I rebased it onto latest develop

miczyg1 commented:
> [fast-boot-tgl.zip](https://github.com/user-attachments/files/19340445/fast-boot-tgl.zip)

Thislog does not match the current code. e.g. the last IF in the log is ` ${relative_boot_time} > 100.0` but should be ` ${relative_boot_time} > 70.0`. Please provide a fresh one

SebastianCzapla commented:
[fast-boot.zip](https://github.com/user-attachments/files/19549319/fast-boot.zip)

SebastianCzapla commented:
[fast-boot.zip](https://github.com/user-attachments/files/19550402/fast-boot.zip)
Log made after rebase and force-push (https://github.com/Dasharo/open-source-firmware-validation/compare/b7d12929e5339ec70627efb2ef7f50923dc71cb4..cba3f247a72547f50dddb325c7d9ed6f62d6bf45)

miczyg1 commented on dasharo-performance/fast-boot.robot:
```diff
@@ -0,0 +1,133 @@
+
+    FOR    ${index}    IN RANGE    0    ${iterations}
+        Power Cycle On
+        Boot System Or From Connected Disk    ${ENV_ID_UBUNTU}
```
This works only because the laptop does not have serial console and does not check for prompt in the firmware (the DUT connection method is SSH on laptops). If one would launch the etst on another machine, it will not work. The platform will boot directly to the OS, so this should be skipped if fast boot is enabled.

SebastianCzapla replied:
Fixed in https://github.com/Dasharo/open-source-firmware-validation/pull/706/commits/0e9d9a8837e8684f10b61c836d1299dc25e1875a

Log from retest of the changes: 
[fast-boot.robot-fix-bootorder.zip](https://github.com/user-attachments/files/19566131/fast-boot.robot-fix-bootorder.zip)

SebastianCzapla committed:
dasharo-performance/fast-boot.robot: Ensure boot priority

Add initialization keyword, in which we setup Ubuntu as first
in list, so that with Fast Boot enabled we always boot the
correct system.

Signed-off-by: Sebastian Czapla <sebastian.czapla@3mdeb.com>

miczyg1 commented on dasharo-performance/fast-boot.robot:
```diff
@@ -0,0 +1,151 @@
+    ...    ${ubuntu_boot_id},${boot_order_no_ubuntu}
+
+    ${out}=    Execute Linux Command    ${set_order_cmd}
+    Should Contain    ${out}    BootCurrent: ${ubuntu_boot_id}
```
BootCurrent says which boot option what been booted currently, so if you change the boot order to ubuntu from ubuntu, itwill always be true. You have to check against BootOrder.

SebastianCzapla replied:
Changed in: https://github.com/Dasharo/open-source-firmware-validation/pull/706/commits/ec65142eac0682d2fac6720dbbb33164bb9b1158

miczyg1 commented on dasharo-performance/fast-boot.robot:
```diff
@@ -0,0 +1,151 @@
+    Should Not Be Empty    ${ubuntu_boot_id}
+
+    ${boot_order_no_ubuntu}=    Execute Linux Command
+    ...    efibootmgr | grep "BootOrder" | awk '{print $2}' | sed -e 's/${ubuntu_boot_id},//g'
```
This will not work if Ubuntu is the last entry in BootOrder. The comma will not be there. So you have three cases:

1. Ubuntu is already first in BootOrder, so skip and do nothing.
2. Ubuntu is in the middle of BootOrder, then sed command will be `sed -e 's/${ubuntu_boot_id},//g'`
3. Ubuntu is last in the BootOrder, then sed command must be `sed -e 's/,${ubuntu_boot_id}//g'` (the comma goes first).

2 and 3 could be merged into  `sed -e 's/,${ubuntu_boot_id}//g'`  because it will work for both the middle and last case. You have to just implement the skip for the case where Ubuntu is already first.

SebastianCzapla replied:
Changed in: https://github.com/Dasharo/open-source-firmware-validation/pull/706/commits/ec65142eac0682d2fac6720dbbb33164bb9b1158

SebastianCzapla replied:
Log of the skipping in action: [fast-boot.robot-skip.zip](https://github.com/user-attachments/files/19570305/fast-boot.robot-skip.zip)

SebastianCzapla replied:
Log for case where Ubuntu is last in the order: [fast-boot.robot-ubuntu-last.zip](https://github.com/user-attachments/files/19570550/fast-boot.robot-ubuntu-last.zip)

miczyg1 commented on dasharo-performance/fast-boot.robot:
```diff
@@ -0,0 +1,151 @@
+    Switch To Root User
+
+    ${ubuntu_boot_id}=    Execute Linux Command
+    ...    efibootmgr | grep -i "ubuntu" | awk '{print $1}' | sed 's/Boot//g' | sed 's/*//g'
```
This can give multiple results, e.g.

`Boot0005* Ubuntu`
`Boot0006* Ubuntu (on <disk>)`

How do you deal with that?

SebastianCzapla replied:
Good call, I don't currently. I will include NR==1 in the awk expression to address that.

SebastianCzapla replied:
Changed in: https://github.com/Dasharo/open-source-firmware-validation/pull/706/commits/ec65142eac0682d2fac6720dbbb33164bb9b1158

SebastianCzapla committed:
dasharo-performance/fast-boot.robot: Fix boot order enforcing

Signed-off-by: Sebastian Czapla <sebastian.czapla@3mdeb.com>

miczyg1 commented on dasharo-performance/fast-boot.robot:
```diff
@@ -0,0 +1,156 @@
+    ${check_if_first_out}=    Execute Linux Command
+    ...    efibootmgr | grep "BootOrder: ${ubuntu_boot_id}" > /dev/null 2>&1 && echo "Correct Order"
+
+    IF    '${check_if_first_out}' != 'Correct Order'
```
```suggestion
    ${check_if_first_out}=    Execute Linux Command
    ...    efibootmgr | grep "BootOrder: ${ubuntu_boot_id}"

    IF    '${check_if_first_out}' == ${EMPTY}
```

Wouldn't this work too?

SebastianCzapla replied:
Changed in https://github.com/Dasharo/open-source-firmware-validation/pull/706/commits/49a7e52bf173eb3aa226a652a7f7eeafedf04232

PR rebased onto today's develop branch

SebastianCzapla replied:
[fast-boot.robot_skip_on_empty.zip](https://github.com/user-attachments/files/19630471/fast-boot.robot_skip_on_empty.zip)

SebastianCzapla committed:
dasharo-performance/fast-boot.robot: Simplify skip condition

Signed-off-by: Sebastian Czapla <sebastian.czapla@3mdeb.com>
